languages/python/django-oso/django_oso/partial.py

```diff
@@ -0,0 +1,99 @@
+from django.db.models import Q
+
+from polar.expression import Expression
+from polar.variable import Variable
+from polar.exceptions import UnsupportedError
+
+
+def partial_to_query_filter(partial, type_name):
+    """
+    Convert a partial expression to a django query ``Q`` object.
+
+    Example expression structure::
+
+        Expression(And, [
+            Expression(Isa, [
+                Variable('_this'),
+                Pattern(test_app::Post, {})]),
+            Expression(Isa, [
+                Variable('_this'),
+                Pattern(test_app::Post, {})]),
+            Expression(Unify, [
+                False,
+                Expression(
+                    Dot, [
+                        Variable('_this'),
+                        'is_private'])])])
+
+    Output::
+
+        Q(is_private=False)
+    """
+
+    q = and_expr(partial, type_name)
+    return q
+
+
+COMPARISONS = {
+    "Unify": lambda q, f, v: Q(**{f: v}),
+    "Eq": lambda q, f, v: Q(**{f: v}),
+    "Neq": lambda q, f, v: ~Q(**{f: v}),
+    "Geq": lambda q, f, v: Q(**{f"{f}__gte": v}),
+    "Gt": lambda q, f, v: Q(**{f"{f}__gt": v}),
+    "Leq": lambda q, f, v: Q(**{f"{f}__leq": v}),
+    "Lt": lambda q, f, v: Q(**{f"{f}__lt": v}),
+}
+
+
+def and_expr(expr, type_name):
+    q = Q()
+
+    assert expr.operator == "And"
+    for expression in expr.args:
+        assert isinstance(expression, Expression)
+        if expression.operator in COMPARISONS:
+            q = q & compare_expr(expression, type_name)
+        elif expression.operator == "And":
+            q = q & and_expr(expression, type_name)
+        elif expression.operator == "Isa":
+            try:
+                assert expression.args[1].tag == type_name
+            except (AssertionError, IndexError, AttributeError, TypeError):
+                raise UnsupportedError(
+                    f"Unimplemented partial isa operation {expression}."
+                )
+        else:
```

Comment:
What about `not` and `or`. Those could make a lot of sense for filtering.

Reply:
Agreed. The semantics of `not` could be difficult, but I agree is worth doing. `or` should be easy, it just hasn't been done yet.

Reply:
`or` actually isn't needed because the results come out as two separate bindings since it generates choice points.